fix(plugin): MCOL-6088 This ensures that `SELECT col AS alias ... GRO…

mariadb-LeonidFedorov opened the pull request:
# Fix for GROUP BY Ambiguous Column Reference Bug

## Problem Description

When a GROUP BY clause contains a column name that is ambiguous (i.e., it matches both a table column and a SELECT list alias pointing to a different column), ColumnStore was throwing an error:

```
ERROR 1178 (42000): The storage engine for the table doesn't support MCS-2016: Non supported item 'cat_id' on the GROUP BY list.
```

The expected behavior is to handle the ambiguity gracefully, similar to how standard MariaDB would handle it (with a warning if appropriate).

## Example Query

```sql
CREATE TABLE test2 (
  id INT UNSIGNED NOT NULL,
  GBP DOUBLE UNSIGNED NOT NULL,
  cat_id INT UNSIGNED NOT NULL,
  cat_id_level_0 INT UNSIGNED NOT NULL
) ENGINE=Columnstore;

-- This query has an ambiguous GROUP BY reference
-- 'cat_id' could refer to:
--   1. The table column 'cat_id'
--   2. The SELECT alias 'cat_id' (which is cat_id_level_0 AS cat_id)

SELECT cat_id_level_0 AS entity_id, SUM(GBP) AS spend, cat_id_level_0 AS cat_id
FROM test2
GROUP BY cat_id;
```

## Root Cause

In the `processGroupBy()` function in `ha_mcs_execplan.cpp`, when processing a `FIELD_ITEM` in the GROUP BY clause:

1. `buildSimpleColumn()` was called to create a column reference
2. If this failed (returned NULL) due to ambiguity or other issues, the code immediately set `nonSupportItem`, triggering the error
3. The code didn't check if there was a matching alias in the SELECT list that could resolve the reference

## Solution

Modified `processGroupBy()` function in `ha_mcs_execplan.cpp` to handle ambiguous column references more gracefully:

1. When `buildSimpleColumn()` succeeds, use the table column and check SELECT list by `sameColumn()` match
2. If table column found in SELECT list, also check for alias match as a fallback
3. When `buildSimpleColumn()` returns NULL or non-SimpleColumn, match by alias name in SELECT list

mariadb-LeonidFedorov committed:
fix(plugin): MCOL-6088 This ensures that `SELECT col AS alias ... GROUP BY alias` works correctly even when `col` is wrapped in an implicit aggregate wrapper.

mariadb-SergeyZefirov commented on dbcon/mysql/ha_mcs_execplan.cpp:
```diff
@@ -5724,6 +5724,61 @@ int processHaving(SELECT_LEX& select_lex, gp_walk_info& gwi, SCSEP& csep,
+    ReturnedColumn* inner_col = agg_check->aggParms()[0].get();
+    AggregateColumn* inner_agg = dynamic_cast<AggregateColumn*>(inner_col);
+
+    if (inner_agg)
```
Can it be recursive? E.g., ``SUM(COUNT(group_by_col))``.

mariadb-SergeyZefirov reviewed (approved):
I agree with changes as the solve exactly what need to be solved and thus can be merged.

I have a question about nested aggregate unwrapping, but I do not see it as super important. If we differ in behavior from the server for single aggregate, we also may differ for nested aggregates.

mariadb-AlekseiBukhalov committed:
MCOL-6088 add tests

mariadb-AlexeyAntipovsky committed:
fix(plugin): fix incorrect behavior for `if` args in case of aggregation

Example:
```sql
create table iftest(a int, b int) engine=columnstore;
insert into iftest values (1, null), (2, 1);
select if (b is null, 'b is null', 'b is not null') from iftest group by a;
```

Incorrect result:
```
ERROR 1815 (HY000): Internal error: MCS-2021: 'if (b is null, 'b is null', 'b is not null')' is not in GROUP BY clause. All non-aggregate columns in the SELECT and ORDER BY clause must be included in the GROUP BY clause.
```

Expected (as with InnoDB):
```
+----------------------------------------------+
| if (b is null, 'b is null', 'b is not null') |
+----------------------------------------------+
| b is null                                    |
| b is not null                                |
+----------------------------------------------+
2 rows in set (0.000 sec)
```

mariadb-LeonidFedorov committed:
fix(joblist): allow functions and arithmetic expressions on GROUP BY columns

When a SELECT contains a function (like length(), upper()) or arithmetic
expression (like length(a) + length(b)) applied to columns that are in
the GROUP BY clause, the query should succeed because the result is
deterministic for each group.

Previously, queries like:
  SELECT length(x) FROM t1 GROUP BY x;
  SELECT length(a) + length(b) FROM t1 GROUP BY a, b;
failed with: "Internal error: 'x' cannot be found in tuple."

Root cause: FunctionColumns and ArithmeticColumns that depend only on
GROUP BY columns were not being recognized as valid non-aggregate
expressions during tuple aggregation. Additionally, simpleColumnList
was not populated before the dependency check.

Fix:
- In doAggProject(): detect FunctionColumns and ArithmeticColumns whose
  simpleColumnList contains only columns present in groupByColVec, and
  add them to groupByColVec so they are treated as GROUP BY expressions
- Call setSimpleColumnList() before checking dependencies to ensure the
  column list is populated
- In tryToFindFunctionDependsOnGroupByColumns(): extend to support
  ArithmeticColumn and use tuple key comparison instead of OID

mariadb-LeonidFedorov committed:
fix(joblist): prevent aggregate function arguments from affecting GROUP BY collation

Commit 682b574c2 introduced a regression where arguments of aggregate
functions (e.g., CONCAT in GROUP_CONCAT) were incorrectly added to
groupByColVec, causing collation-aware grouping to fail.

Problem:
In doAggProject(), the loop over retCols appends GROUP_CONCAT arguments
to retCols (lines 844-856). Later in the same loop, these appended
columns were processed by checkAndAddExpressionDependingOnGroupBy(),
adding them to groupByColVec. This caused grouping to occur by the
function result (binary comparison) instead of the original column
with its collation.

Example of broken behavior:
  SELECT GROUP_CONCAT(CONCAT('[',c1,']')) FROM t1 GROUP BY c1;
  -- With latin1_swedish_ci: 'A','a','Ã','ã' were treated as 6 separate
  -- groups instead of being grouped together.

Fix:
Track the original size of retCols before the loop and only call
checkAndAddExpressionDependingOnGroupBy() for original projection
columns (i < originalRetColsSize), not for columns added from
aggregate function arguments.

This preserves the fix from 682b574c2 for queries like:
  SELECT length(x) FROM t1 GROUP BY x;
while fixing the regression for GROUP_CONCAT and similar aggregates.

Fixes 19 failing ctype_cmp_* MTR tests.

mariadb-LeonidFedorov committed:
fix(joblist): do not add corellated to the same pile

mariadb-LeonidFedorov committed:
fix(joblist): implements a different approach to handle functions that depend only on GROUP BY columns
(like length(x) when x is in GROUP BY). Instead of adding such functions to groupByColVec, evaluate
them after aggregation using the existing prepExpressionOnAggregate mechanism.

drrtuy commented on mysql-test/columnstore/basic/t/mcol-4525.test:
```diff
@@ -83,11 +83,10 @@ INSERT INTO t1 VALUES(10,'bye',10);
 SELECT col1 c FROM t1 ORDER BY AVG(col1);
 SET columnstore_select_handler=AUTO;
-SELECT SUM(col1) AS col2 FROM t1 GROUP BY col2;
+SELECT SUM(col1) AS col2 FROM t1 GROUP BY col2 ORDER BY col2;
```
I suggest to use sorted_result instead of OB if OB is not needed.

drrtuy commented on mysql-test/columnstore/basic/t/mcol-4525.test:
```diff
@@ -83,11 +83,10 @@ INSERT INTO t1 VALUES(10,'bye',10);
 SET columnstore_select_handler=ON;
---error 1178
-SELECT SUM(col1) AS col2 FROM t1 GROUP BY col2;
+SELECT SUM(col1) AS col2 FROM t1 GROUP BY col2 ORDER BY col2;
```
I suggest to use sorted_result instead of OB if OB is not needed.

mariadb-LeonidFedorov committed:
fix(joblist): prevent GROUP_CONCAT arguments from being added to expressionVec

The previous commit (388da94b4) introduced logic to add expressions
depending only on GROUP BY columns to expressionVec for post-aggregation
evaluation. However, this incorrectly included function arguments from
GROUP_CONCAT (e.g., CONCAT('[',c1,']') in GROUP_CONCAT(CONCAT(...))),
causing these functions to return NULL.

Changes:
- Add originalRetColsSize to track original projection columns vs
  columns added later from aggregate function arguments (GROUP_CONCAT)
- Only check for GROUP BY-dependent expressions within original
  projection columns, not aggregate function arguments
- Use getTupleKey with add=false to avoid modifying jobInfo state,
  which could cause issues with cursor re-execution

mariadb-LeonidFedorov committed:
fix(joblist): exclude conditional functions from post-aggregation evaluation

The previous commit added logic to evaluate functions depending only on
GROUP BY columns after aggregation (e.g., length(x) GROUP BY x). However,
this incorrectly included conditional functions like IF(), CASE, IFNULL,
NULLIF, COALESCE, etc., causing them to return wrong results.

For example, IF(value<1000,'<1000','>=1000') GROUP BY value was returning
'<1000' for all rows instead of evaluating the condition per-row.

Changes:
- Add a blacklist of conditional/non-deterministic functions that should
  NOT be evaluated after aggregation: if, case, ifnull, nullif, coalesce,
  isnull, nvl, nvl2, greatest, least, elt, field, find_in_set, make_set,
  rand, uuid, now, curdate, curtime, sysdate, unix_timestamp
- Only pure functions (length, upper, lower, concat, etc.) are now added
  to expressionVec for post-aggregation evaluation
- Add comprehensive tests for conditional functions with GROUP BY
